### PR TITLE
Reset the internal state of the embedded object manager to its previous state

### LINKED ISSUE #4018: Nested HtmlOutputProcessors with embedded objects modify the internal state
This is kind of an obscure bug that just recently surfaced in combination with the bbcodes for threads and posts. It was observed that multiple of such bbcodes within one message will cause only the first one to be rendered. Actually, depending on the circumstances, additional bbcodes could be rendered, if they are embedded objects of the first referenced object.

The issue unfolds like this:
1. A message with support for embedded objects is being rendered.
2. `HtmlOutputProcessor::setContext()` sets the active message of `MessageEmbeddedObjectManager`.
3. During processing, a bbcode is encounted that makes use of `HtmlOutputProcessor` itself (here: Render an excerpt of a post).
4. The same logic as in step 2 happens, setting the internal state of the `MessageEmbeddedObjectManager` to the object in step 3.
5. Once the embedded message is rendered, the control flow switches back to the original message.
6. The next bbcode in this message attempts to read from the `MessageEmbeddedObjectManager`, but the active message points to the embedded message in step 3 now.
7. The bbcode is unable to render the data and its fallback logic kicks in.

### PR DESCRIPTION
**Pull request by dtdesign**
Fixes #4018

### CODE REVIEW

File: wcfsetup/install/files/lib/system/html/output/HtmlOutputProcessor.class.php [hunk 1]
```diff
@@ -90,7 +90,11 @@ public function setOutputType($outputType) {
 	 * @inheritDoc
 	 */
 	public function getHtml() {
-		return $this->getHtmlOutputNodeProcessor()->getHtml();
+		$html = $this->getHtmlOutputNodeProcessor()->getHtml();
+
+		MessageEmbeddedObjectManager::getInstance()->reset();
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/html/output/HtmlOutputProcessor.class.php:95]: Will this cause issues if `->getHtml()` is being called more than once for a single message?

File: wcfsetup/install/files/lib/system/message/embedded/object/MessageEmbeddedObjectManager.class.php [hunk 2]
```diff
@@ -66,6 +66,12 @@ class MessageEmbeddedObjectManager extends SingletonFactory {
 		'insert' => [],
 		'remove' => []
 	];
+
+	/**
+	 * A list of previous active message settings used to restore
+	 * the internal state in case of nested message processing.
+	 */
+	protected $acitveMessageHistory = [];
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/message/embedded/object/MessageEmbeddedObjectManager.class.php:74]:
```suggestion
	protected $activeMessageHistory = [];
```
(adjust the other location yourself, please).

File: wcfsetup/install/files/lib/system/message/embedded/object/MessageEmbeddedObjectManager.class.php [hunk 3]
```diff
@@ -283,10 +289,36 @@ public function getContentLanguageID() {
...
 		$this->activeMessageObjectTypeID = ObjectTypeCache::getInstance()->getObjectTypeIDByName('com.woltlab.wcf.message', $messageObjectType);
 		$this->activeMessageID = $messageID;
 		$this->activeMessageLanguageID = $languageID;
 	}
+
+	/**
+	 * Restores the internal state in case of nested message processing.
+	 */
+	public function reset() {
+		$newState = array_pop($this->acitveMessageHistory);
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/message/embedded/object/MessageEmbeddedObjectManager.class.php:309]: When merging this up: Don't forget to add the `\` (or simply also add it for 5.3).

**Review by TimWolla — CHANGES_REQUESTED** [review]
Overall this looks good to me, just a few small nits.

File: wcfsetup/install/files/lib/system/html/output/HtmlOutputProcessor.class.php (hunk 1 above)
Comment (dtdesign) [reply wcfsetup/install/files/lib/system/html/output/HtmlOutputProcessor.class.php:95]: Good catch, this would indeed cause issues. Duh.

**Review by TimWolla — APPROVED** [review]

**Pull request merged**